common/chat.cpp

```diff
@@ -2083,63 +2083,71 @@ static void common_chat_parse_granite(common_chat_msg_parser & builder) {
 static common_chat_params common_chat_params_init_qwen3_coder_xml(const common_chat_template & tmpl, const struct templates_params & inputs) {
     common_chat_params data;
     data.grammar_lazy = inputs.tool_choice != COMMON_CHAT_TOOL_CHOICE_REQUIRED;
-    
+
     // Always set the format to QWEN3_CODER_XML regardless of whether tools are provided
     // The format identifies the template type, not the runtime configuration
     data.format = COMMON_CHAT_FORMAT_QWEN3_CODER_XML;
-    
+
     if (!inputs.tools.empty()) {
         data.grammar = build_grammar([&](const common_grammar_builder & builder) {
             std::vector<std::string> tool_rules;
-            std::vector<std::string> escaped_names;
-            
+
+            auto not_parameter_end = builder.add_rule("not_parameter_end", "([^<] | (\"<\" [^/]) | (\"</\" [^p]) | (\"</p\" [^a]) | (\"</pa\" [^r]) | (\"</par\" [^a]) | (\"</para\" [^m]) | (\"</param\" [^e]) | (\"</parame\" [^t]) | (\"</paramet\" [^e]) | (\"</paramete\" [^r]) | (\"</parameter\" [^>]))*");
+
             foreach_function(inputs.tools, [&](const json & tool) {
                 const auto & function = tool.at("function");
                 const std::string & name = function.at("name");
                 const json & parameters = function.at("parameters");
-                
-                escaped_names.push_back(regex_escape(name));
-                
+
+                std::unordered_set<std::string> required;
+                if (parameters.contains("required")) {
+                    for (const auto & p : parameters.at("required")) {
+                        required.insert(p);
+                    }
+                }
+
                 // Build parameter rules for XML format
                 std::vector<std::string> param_rules;
                 if (parameters.contains("properties")) {
                     for (const auto & [param_name, param_schema] : parameters["properties"].items()) {
                         std::string param_rule = "\"<parameter=" + param_name + ">\" space ";
-                        
+
                         // Add parameter value based on type
                         if (param_schema.contains("type")) {
                             std::string param_type = param_schema["type"];
                             if (param_type == "string") {
-                                param_rule += "[^<]* ";
-                            } else if (param_type == "integer" || param_type == "number") {
-                                param_rule += "[0-9.-]+ ";
-                            } else if (param_type == "boolean") {
-                                param_rule += "(\"true\" | \"false\") ";
+                                param_rule += not_parameter_end + " ";
```

Comment:
Using builder.add_schema here seemed to cause the model to generate a single line in a file and then do an edit to add the rest. Kinda weird so went with my original "fix" here to except anything except `</parameter>`.